internal/http_client.go

```diff
@@ -0,0 +1,193 @@
+
+package internal
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"golang.org/x/net/context"
+)
+
+// HTTPClient is a convenient API to make HTTP calls.
+//
+// This API handles some of the repetitive tasks such as entity serialization and deserialization
+// involved in making HTTP calls. It provides a convenient mechanism to set headers and query
+// parameters on outgoing requests, while enforcing that an explicit context is used per request.
+// Responses returned by HTTPClient can be easily parsed as JSON, and provide a simple mechanism to
+// extract error details.
+type HTTPClient struct {
+	Client    *http.Client
+	ErrParser ErrorParser
+}
+
+// Do executes the given Request, and returns a Response.
+func (c *HTTPClient) Do(ctx context.Context, r *Request) (*Response, error) {
+	req, err := r.newHTTPRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.Client.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return &Response{
+		Status:    resp.StatusCode,
+		Body:      b,
+		Header:    resp.Header,
+		errParser: c.ErrParser,
+	}, nil
+}
+
+// Request contains all the parameters required to construct an outgoing HTTP request.
+type Request struct {
+	Method string
+	URL    string
+	Body   HTTPEntity
+	Opts   []HTTPOption
+}
+
+func (r *Request) newHTTPRequest() (*http.Request, error) {
+	var opts []HTTPOption
+	var data io.Reader
+	if r.Body != nil {
+		b, err := r.Body.Bytes()
+		if err != nil {
+			return nil, err
+		}
+		data = bytes.NewBuffer(b)
+		opts = append(opts, WithHeader("Content-Type", r.Body.Mime()))
+	}
+
+	req, err := http.NewRequest(r.Method, r.URL, data)
+	if err != nil {
+		return nil, err
+	}
+
+	opts = append(opts, r.Opts...)
+	for _, o := range opts {
+		o(req)
+	}
+	return req, nil
+}
+
+// HTTPEntity represents a payload that can be included in an outgoing HTTP request.
+type HTTPEntity interface {
+	Bytes() ([]byte, error)
+	Mime() string
+}
+
+type jsonEntity struct {
+	Val interface{}
+}
+
+// NewJSONEntity creates a new HTTPEntity that will be serialized into JSON.
+func NewJSONEntity(v interface{}) HTTPEntity {
+	return &jsonEntity{Val: v}
+}
+
+func (e *jsonEntity) Bytes() ([]byte, error) {
+	return json.Marshal(e.Val)
+}
+
+func (e *jsonEntity) Mime() string {
+	return "application/json"
+}
+
+// Response contains information extracted from an HTTP response.
+type Response struct {
+	Status    int
+	Header    http.Header
+	Body      []byte
+	errParser ErrorParser
+}
+
+// CheckStatus checks whether the Response status code has the given HTTP status code.
+//
+// Returns an error if the status code does not match. If an ErroParser is specified, uses that to
+// construct the returned error message. Otherwise includes the full response body in the error.
+func (r *Response) CheckStatus(want int) error {
+	if r.Status == want {
```

Comment:
One concern I have about this is that there are HTTP endpoints where any 200 response is acceptable, and the client may not care at all if it's 200 or 201 or whatever. Are there cases where this CheckStatus actually needs to check some other code than a 200?

Reply:
There are cases in RTDB where we would check for 2 status codes (e.g. 200 and 412 from a conditional request), and handle the response accordingly. Something like:

```
if err := resp.CheckStatus(200); err == nil {
  // Operation successful -- Parse response
} else if err := resp.CheckStatus(412); err == nil {
  // Expected failure -- Parse error response accordingly
} else {
  // Unexpected failure
  return err
}
```